### PR TITLE
Refactor NMS getParticle() implementation to be easier to maintain

### PR DESCRIPTION
**Pull request by AkaGiant**
Please ignore the fact it took me three commits to figure it out.

### CODE REVIEW

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 1]
```diff
@@ -160,6 +160,8 @@
 import net.minecraft.world.level.gameevent.EntityPositionSource;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:164]: Unnecessary imports that aren't useful for anyone other than maintainers. The methods these are used on are only being used internally by the CommandAPI and aren't publicly seen by end users (main developers), so they have no main purpose.

In general, almost all methods in the CommandAPI are non-null, so these are effectively obsolete here.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 2]
```diff
@@ -577,6 +536,111 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
 		return new ParticleData<Void>(particle, null);
 	}
 
+	/**
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:539]: JavaDocs with no Javadocs! You can't have JavaDocs with no description for the parameters!

In this case however, we don't need JavaDocs here because this is a `private` method.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 3]
```diff
@@ -577,6 +536,111 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
...
+		return new ParticleData<>(particle, new DustTransition(color, toColor, options.getScale()));
+	}
+
+	/**
+	 * {@link #getParticle(CommandContext, String)}  Parent Method}
+	 * @param particle
+	 * @param options
+	 * @return
+	 */
+	private ParticleData<DustOptions> getParticleDataAsDustParticleOptions(Particle particle, DustParticleOptions options) {
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:559]: This method is really small, did it really need its own method for these two statements?

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 4]
```diff
@@ -577,6 +536,111 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
...
+	}
+	
+	/**
+	 * {@link #getParticleDataAsVibrationParticleOption(CommandContext, Level, Particle, VibrationParticleOption)
+	 * Parent Method}
+	 * @param positionSource {@link BlockPositionSource} the block position source to find the designation of.
+	 * @param level {@link Level} The level in which to find the destination
+	 * @return {@link Destination} the destination of the provided source in the given level.
+	 */
+	private @NotNull Destination getVibrationParticleOptionDestinationAsBlockPositionSource(@NotNull BlockPositionSource positionSource, Level level) {
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:600]: This method is really small, did it really need its own method for these two statements?

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 5]
```diff
@@ -726,12 +791,7 @@ public final void reloadDataPacks() {
 			List<String> enabledIDs = ImmutableList.copyOf(selectedIDs);
 			List<String> disabledIDs = new ArrayList<>(MINECRAFT_SERVER.getPackRepository().getAvailableIds());
 
-			ListIterator<String> disabledIDsIterator = disabledIDs.listIterator();
-			while (disabledIDsIterator.hasNext()) {
-				if (enabledIDs.contains(disabledIDsIterator.next())) {
-					disabledIDsIterator.remove();
-				}
-			}
+			disabledIDs.removeIf(enabledIDs::contains);
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:794]: Woah, this is a really nice way of doing this. Awesome! This should mean that the `ListIterator` import is no longer required anymore, so that should be removed.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 6]
```diff
@@ -714,7 +779,7 @@ public final void reloadDataPacks() {
 			return simpleReloadInstance.thenApply(x -> serverResources);
 		});
 
-		// Step 3: Actually load all of the resources
+		// Step 3: Actually load all the resources
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:782]: Unnecessary change here, please revert.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 7]
```diff
@@ -594,6 +658,7 @@ public CommandSender getSenderForCommand(CommandContext<CommandSourceStack> cmdC
 
 		Entity proxyEntity = css.getEntity();
 		CommandSender proxy = proxyEntity == null ? null : proxyEntity.getBukkitEntity();
+
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:661]: Unnecessary new line, please revert. I have a series of diffing tools that I use to compare NMS versions and this will trip it up.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 8]
```diff
@@ -577,6 +536,111 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
...
+		assert MINECRAFT_SERVER != null;
+		for (ServerLevel world : MINECRAFT_SERVER.getAllLevels()) {
+			Entity entityById = world.getEntity(id.right().get());
+			if (entityById != null) {
+				foundEntity = entityById;
+				break;
+			}
+		}
+		// Asset found entity is not null to stop null pointer exception.
+		assert foundEntity != null;
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:640]: We can assume `foundEntity` is not null. This is actually checked in Minecraft's game packets when it triggers a vibration particle. In the rare case that it doesn't catch this, an exception will be thrown and caught and shown to users (that something went wrong when trying to input that particle argument), which means that they probably entered the wrong entity ID or something which is perfectly fine.

Please remove this assertion.

**Review by JorelAli — CHANGES_REQUESTED** [review]
Nice PR! The changes make the code much more readable and easier to understand.

There are a few changes that I would like and have mentioned these in this review.

The changes in this PR only affect `NMS_1_19_Common`. There are also a few other classes which have this `getParticle()` method which should also be refactored in the same way:
- `NMS_1_17_Common`
- `NMS_1_18_R1`
- `NMS_1_18_R2`
- `NMS_1_19_3_R2`

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 9]
```diff
@@ -510,60 +512,17 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
 			return new ParticleData<BlockData>(particle, CraftBlockData.fromData(options.getState()));
 		}
 		if (particleOptions instanceof DustColorTransitionOptions options) {
```
[the remark this answers is not in the document]
Comment (AkaGiant) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:514]: Could you explain why? Personally i see that as a more preference but would love to get a better understanding if there's more of a practical usage :)

[the remark this answers is not in the document]
Comment (JorelAli) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:514]: It's safer. It's very easy to make the mistake of something like this:

```java
if () {
  // case 1
  return ...
}
if () {
  // case 2
  return ...
}
if () {
  // case 3
  
}
if () {
  // case 4
  return ...
}
// Cases 6+
return ...
```

In this scenario, we've missed a return on case 3, and it now "falls through" the logic to cases 6+. This doesn't occur with an else statement:

```java
if () {
  // case 1
  return ...
} else if () {
  // case 2
  return ...
} else if () {
  // case 3
  
} else if () {
  // case 4
  return ...
} else {
  // Cases 6+
  return ...
}
// Compiler error - 3 didn't return anything
```

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java (hunk 6 above)
Comment (AkaGiant) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:782]: is this regarding the "of" ?

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java (hunk 5 above)
Comment (AkaGiant) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:794]: Thanks, ive also removed that. :)

**Review by AkaGiant — COMMENTED** [review]
So now how do I like add the new version?

**Comment by AkaGiant** [comment]
I'll go through and do the other files too but how do I "update" it with my changes? open a new PR?

**Comment by JorelAli** [comment]
image

**Comment by AkaGiant** [comment]
oh damn i didn't realise it would just work like that

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 10]
```diff
@@ -503,45 +468,41 @@ public final org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSource
...
+		else if (particleOptions instanceof ShriekParticleOption options) {
 			// CraftBukkit implements shriek particles as a (boxed) Integer object
 			return new ParticleData<Integer>(particle, Integer.valueOf(options.getDelay()));
 		}
-		if (particleOptions instanceof SculkChargeParticleOptions options) {
+		else if (particleOptions instanceof SculkChargeParticleOptions options) {
 			// CraftBukkit implements sculk charge particles as a (boxed) Float object
 			return new ParticleData<Float>(particle, Float.valueOf(options.roll()));
 		}
 		CommandAPI.getLogger().warning("Invalid particle data type for " + particle.getDataType().toString());
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:502]: This last part should be inside an `else` statement

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 11]
```diff
@@ -503,45 +468,41 @@ public final org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSource
 	public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdCtx, String str) {
 		final ParticleOptions particleOptions = ParticleArgument.getParticle(cmdCtx, str);
 		final Level level = cmdCtx.getSource().getLevel();
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:470]: This `level` variable doesn't get used by anything other than `VibrationParticleOption`. We can remove this and use:

```java
return getParticleDataAsVibrationParticleOption(cmdCtx, cmdCtx.getSource().getLevel(), particle, options);
```

**Review by JorelAli — CHANGES_REQUESTED** [review]

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 12]
```diff
@@ -467,7 +467,6 @@ public final org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSource
 	@Override
 	public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdCtx, String str) {
 		final ParticleOptions particleOptions = ParticleArgument.getParticle(cmdCtx, str);
-		final Level level = cmdCtx.getSource().getLevel();
 		final Particle particle = CraftParticle.toBukkit(particleOptions);
 
 		if (particleOptions instanceof SimpleParticleType options) {
```
[the remark this answers is not in the document]
Comment (JorelAli) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java:472]: Please remove it then:
```java
if (particleOptions instanceof SimpleParticleType) {
  // ...
```

**Comment by JorelAli** [comment]
Looks great! Please implement for the other NMS implementations.

File: commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java [hunk 13]
```diff
@@ -489,7 +488,7 @@ else if (particleOptions instanceof ItemParticleOption options) {
 				CraftItemStack.asBukkitCopy(options.getItem()));
 		}
 		else if (particleOptions instanceof VibrationParticleOption options) {
-			return getParticleDataAsVibrationParticleOption(cmdCtx, level, particle, options);
```
[the remark this answers is not in the document]
Comment (JorelAli) [reply commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java]: I noticed, thanks very much. Looks good.

**Comment by AkaGiant** [comment]
Sweet! thanks for changing the title as I had no idea what to name it :)

**Review by AkaGiant — COMMENTED** [review]
I'm almost 100% certain this was missed during the clean up of the previous few commits.

File: commandapi-nms/commandapi-1.18.2/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R2.java [hunk 14]
```diff
@@ -484,52 +453,65 @@ public org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSourceStack>
...
 		}
-		if (particleOptions instanceof DustParticleOptions options) {
+		else if (particleOptions instanceof DustParticleOptions options) {
 			final Color color = Color.fromRGB((int) (options.getColor().x() * 255.0F),
 					(int) (options.getColor().y() * 255.0F), (int) (options.getColor().z() * 255.0F));
-			return new ParticleData<DustOptions>(particle, new DustOptions(color, options.getScale()));
+			return new ParticleData<>(particle, new DustOptions(color, options.getScale()));
 		}
-		if (particleOptions instanceof ItemParticleOption options) {
-			return new ParticleData<org.bukkit.inventory.ItemStack>(particle,
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.18.2/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R2.java:509]: I want these explicit types (`org.bukkit.inventory.ItemStack` and `DustOptions`) back

File: commandapi-nms/commandapi-1.18/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R1.java [hunk 15]
```diff
@@ -433,52 +429,65 @@ public org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSourceStack>
...
 		}
-		if (particleOptions instanceof DustParticleOptions options) {
+		else if (particleOptions instanceof DustParticleOptions options) {
 			final Color color = Color.fromRGB((int) (options.getColor().x() * 255.0F),
 					(int) (options.getColor().y() * 255.0F), (int) (options.getColor().z() * 255.0F));
-			return new ParticleData<DustOptions>(particle, new DustOptions(color, options.getScale()));
+			return new ParticleData<>(particle, new DustOptions(color, options.getScale()));
 		}
-		if (particleOptions instanceof ItemParticleOption options) {
-			return new ParticleData<org.bukkit.inventory.ItemStack>(particle,
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.18/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R1.java:458]: I want these explicit types (`org.bukkit.inventory.ItemStack` and `DustOptions`) back

File: commandapi-nms/commandapi-1.18/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R1.java [hunk 16]
```diff
@@ -433,52 +429,65 @@ public org.bukkit.loot.LootTable getLootTable(CommandContext<CommandSourceStack>
...
+			destination = new EntityDestination(entity.get().getBukkitEntity());
+		}
+		else {
+			CommandAPI.getLogger()
+				.warning("Unknown vibration destination " + options.getVibrationPath().getDestination());
+			return new ParticleData<Void>(particle, null);
 		}
-		CommandAPI.getLogger().warning("Invalid particle data type for " + particle.getDataType().toString());
-		return new ParticleData<Void>(particle, null);
+		return new ParticleData<>(particle,
```
Comment (JorelAli) [inline commandapi-nms/commandapi-1.18/src/main/java/dev/jorel/commandapi/nms/NMS_1_18_R1.java:489]: I want an explicit generic type here (`Vibration`)

**Review by JorelAli — CHANGES_REQUESTED** [review]

**Review by JorelAli — APPROVED** [review]
